# Quiz1 - Numpy
## Submitted by: Rica Eunice L. Dordas 

#### Part 1.

In [40]:
import numpy as np

NumPy Package will be imported and referred to as np.

In [41]:
y_hat = np.linspace(-1, 1, num=20)
y_hat

array([-1.        , -0.89473684, -0.78947368, -0.68421053, -0.57894737,
       -0.47368421, -0.36842105, -0.26315789, -0.15789474, -0.05263158,
        0.05263158,  0.15789474,  0.26315789,  0.36842105,  0.47368421,
        0.57894737,  0.68421053,  0.78947368,  0.89473684,  1.        ])

y_hat is an initialized array with 20 linearly spaced numbers between -1 and 1.

In [42]:
y = np.linspace(0, 1, num= 20)
y

array([0.        , 0.05263158, 0.10526316, 0.15789474, 0.21052632,
       0.26315789, 0.31578947, 0.36842105, 0.42105263, 0.47368421,
       0.52631579, 0.57894737, 0.63157895, 0.68421053, 0.73684211,
       0.78947368, 0.84210526, 0.89473684, 0.94736842, 1.        ])

y is an initialized array with 20 linearly spaced numbers between 0 and 1.

In [43]:
y_hat.shape

(20,)

y_hat has 20 rows with only 1 column.

In [44]:
y.shape

(20,)

y has 20 rows with only 1 column.

In [45]:
y_hat.ndim

1

y_hat has a dimension of 1 which states that it is a 1D array.

In [46]:
y.ndim

1

y has a dimension of 1 which states that it is a 1D array.

In [47]:
def MSE (y, y_hat):
    return np.square(y-y_hat).mean()

MSE(y,y_hat)

0.3421052631578948

The Mean Squared Error with y and y_hat as parameters is 0.3421052631578948.

In [48]:
def MAE (y, y_hat):
    return np.abs(y_hat-y).mean()

MAE(y,y_hat)

0.4999999999999999

The Mean Absolute Error with y and y_hat as parameters is 0.4999999999999999.

In [49]:
def MSLE (y, y_hat):
    return np.square((np.log(y_hat+1)-np.log(y+1)))

MSLE(y,y_hat)

C:\Users\Rica Eunice\AppData\Local\Temp\ipykernel_23104\1654995694.py:2: RuntimeWarning: divide by zero encountered in log
  return np.square((np.log(y_hat+1)-np.log(y+1)))


array([           inf, 5.30189811e+00, 2.74972035e+00, 1.68813627e+00,
       1.11524725e+00, 7.66445510e-01, 5.38710750e-01, 3.83209542e-01,
       2.73788620e-01, 1.95216181e-01, 1.38059476e-01, 9.61960796e-02,
       6.55018920e-02, 4.31141058e-02, 2.69954927e-02, 1.56658124e-02,
       8.03033899e-03, 3.26708427e-03, 7.50703787e-04, 0.00000000e+00])

The Mean Squared Logarithmic Error with y and y_hat as parameters cannot be computed since y has the value of 0 which leads to an error since it cannot be a divisor.

In [50]:
def MAPE (y, y_hat):
    return (np.abs((y_hat-y)/y).mean())*100

MAPE(y,y_hat)

C:\Users\Rica Eunice\AppData\Local\Temp\ipykernel_23104\1419580328.py:2: RuntimeWarning: divide by zero encountered in true_divide
  return (np.abs((y_hat-y)/y).mean())*100


inf

The Mean Absolute Percentage Error with y and y_hat as parameters cannot be computed since y has the value of 0 which leads to an error since it cannot be a divisor.

#### Part 2.

In [51]:
M = np.arange(1, 17)
M

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16])

arange function is used to create an array ranges 1-16.

In [52]:
N = M.reshape(4,4)
N

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

reshape array is used to change the shape of an array. In this number, since the array has 16 values, it can be equally reshaped to 4 rows and 4 columns.

In [53]:
N[1]

array([5, 6, 7, 8])

N[i] where i is the index of the row you want to get the values from. Since we want to get the 2nd row, we will use N[1] where 1 is the index of the 2nd row.

In [54]:
np.min(N, axis=0)

array([1, 2, 3, 4])

min function is used to get the miniumum value for each column/row. Since we need to find the min value for each column, we will assign the axis to 0. The min values of each column are 1, 2, 3, and 4. 

In [55]:
np.argmin(N, axis=0)

array([0, 0, 0, 0], dtype=int64)

argmin is used to find the location which is the index of the min values. Since all min values for each column are located in the 1st row, their location is 0.

In [56]:
np.mean(N)

8.5

mean function is used to find the average of the whole matrix. For matrix N, the mean is 8.5.

In [57]:
V = np.array([1,1,1,1])
V

array([1, 1, 1, 1])

A vector can be created through a 1d array. Vector V was created with the 4 values which are all 1.

In [58]:
np.dot(N, V)

array([10, 26, 42, 58])

For the dot product of a matrix and vector, each row of the matrix will be multiplied to a corresponding value in the vector then get the sum of all the product. For example the values of the 1st row of the matrix are 1, 2, 3, and 4. These values will all be multiplied to 1 since these are the values of the vector. Then, all of the products will be added.